zk-token-sdk/src/zk_token_elgamal/pod/auth_encryption.rs

```diff
@@ -34,6 +44,25 @@ impl fmt::Display for AeCiphertext {
     }
 }
 
+#[cfg(not(target_os = "solana"))]
+impl FromStr for AeCiphertext {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() > AE_CIPHERTEXT_MAX_BASE64_LEN {
+            return Err(ParseError::WrongSize);
+        }
+        let ciphertext_vec = BASE64_STANDARD.decode(s).map_err(|_| ParseError::Invalid)?;
```

Comment:
How about decoding directly into an array of `[u8; AE_CIPHERTEXT_LEN]` with `decode_slice` https://docs.rs/base64/latest/base64/engine/trait.Engine.html#method.decode_slice? Then you can check that the length is correct and return that directly, no allocations required!

Reply:
This is a good idea, but one issue that I had is that something like this does not work
```
let mut array = [0u8; AE_CIPHERTEXT_LEN]; // 64 byte array
let decoded_length = BASE64_STANDARD.decode_slice(s, &mut array).unwrap();
```
because it seems like for the base64 encoding, you need extra space for the additional zeroes during decoding. So `decoded_length` is still 64, but `array` must be at least of size `66` for this to not panic. This seems to make directly decoding into an array pointless since we would need to eventually allocate to an array length 64. What do you think?

Reply:
Sorry, I think I'm missing something, do you have an example that shows this? I just tried this with a few examples of valid of 64-byte encoded strings, ie:

```
    let s = "VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIDFUaGUgcXVpY2sgYnJvd24gZm94IGp1bXBzIG92ZXIgMQ==";
    let mut array = [0u8; 64];
    let decoded_length = BASE64_STANDARD.decode_slice(&s, &mut array).unwrap();
    println!("decoded: {decoded_length}, array: {array:?}");
```

And it worked as expected with a decoded length of 64.

Reply:
Hm, weird... I just copy pasted the exact lines into a temporary test and it seems to give the following output for me:
```
thread 'zk_token_elgamal::pod::elgamal::tests::temp_test' panicked at zk-token-sdk/src/zk_token_elgamal/pod/elgamal.rs:186:75:
called `Result::unwrap()` on an `Err` value: OutputSliceTooSmall
```

If you add the following test
```
    #[test]
    fn temp_test() {
        let s = "VGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIDFUaGUgcXVpY2sgYnJvd24gZm94IGp1bXBzIG92ZXIgMQ==";
        let mut array = [0u8; 64];
        let decoded_length = BASE64_STANDARD.decode_slice(&s, &mut array).unwrap();
        println!("decoded: {decoded_length}, array: {array:?}");
    }
```
does it pass in your local machine? Both my local and dev server seems to produce `OutputSliceTooSmall`.

Reply:
Oh! Looks like this was fixed in version 0.22, which is the version I was testing with

Reply:
https://github.com/marshallpierce/rust-base64/blob/master/RELEASE-NOTES.md#0220 literally the exact situation we're discussing :sweat_smile:

Reply:
#225 on it!